### PR TITLE
VIH-8043 service web log off page not found

### PR DESCRIPTION
**Pull request by MichaelKeaneathmcts**
### JIRA link ###
https://tools.hmcts.net/jira/browse/VIH-8043

### Change description ###

Fix for application changes

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by ibthmcts — DISMISSED** [review]

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html [hunk 1]
```diff
@@ -17,8 +17,8 @@
       <nav class="hmcts-header__navigation" aria-label="Account navigation">
         <ul class="hmcts-header__navigation-list">
           <li *ngIf="loggedIn" class="hmcts-header__navigation-item">
-            <a id="header-logout-link" class="govuk-header__link  right-align" href="javascript:void(0)"
-              (click)="logout()" i18n="@@header_btn_signout">
+            <a id="header-logout-link" class="govuk-header__link  right-align" href="/logout"
+               i18n="@@header_btn_signout">
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: I'm not sure we should use href in an angular app as it reloads the page. It might be fine since it's a logout but I will have to confirm.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts [hunk 2]
```diff
@@ -56,9 +56,9 @@ describe('HeaderComponent',
...
       fixture.detectChanges();
 
-      const link = fixture.debugElement.query(By.css('#header-logout-link'));
-      link.triggerEventHandler('click', null);
-
-      expect(router.navigate).toHaveBeenCalledWith(['/logout']);
+      const signOutElement = fixture.debugElement.queryAll(By.css('#header-logout-link'));
+      expect(signOutElement.length).toBeGreaterThan(0);
+      const el = signOutElement[0].nativeElement as HTMLElement;
+      expect(el.innerHTML).toContain('Sign out');
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:62]: This only checks the content and not the action of clicking.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html (hunk 1 above)
Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: This could cause issues as the logic is in the logout component. Please revert or use routerLink instead.

**Review by FarronF — CHANGES_REQUESTED** [review]
Please see comments

**Comment by MichaelKeaneathmcts** [comment]
@krishnapottekkat-hmcts Can you look at the comments and set a resolution.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html (hunk 1 above)
Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: https://angular.io/api/router/RouterLink

Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: A similar change the other way being made in video web: https://github.com/hmcts/vh-video-web/pull/1510/files#r686123961

We should make sure we have a consistent way to do this.

Comment (krishnapottekkat-hmcts) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: The previous javascript function was having an erroneous effect and made it compatible with the usage and pattern followed elsewhere  in Video Web and Admin Web. In that case, we need to standardise this as a technical debt and need to standardise in all the solutions

Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: Going forward we will use the routerlink method seen in the PR in my previous comment. We should make that change now rather than waiting.

Comment (MichaelKeaneathmcts) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.html:21]: Changes here may break automation, please keep us up to speed so we can change our testing at the earliest opportunity.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts [hunk 3]
```diff
@@ -1,64 +1,50 @@
...
-      return Promise.resolve(profile);
-  }
-
-  public get isLoggedIn(): boolean {
-      return true;
-  }
+    getUserProfile(): Promise<UserProfile> {
+        const profile = new UserProfile();
+        profile.role = 'professional';
+        profile.email = 'professional@hearings.hmcts.net';
```
Comment (alberto-hortiguela) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:12]: @hearings.hmcts.net email accounts might be flagged by the security team, can we just use @hmcts.net instead?

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts [hunk 4]
```diff
@@ -1,64 +1,50 @@
...
-      fixture.detectChanges();
+        component.loggedIn = true;
+        fixture.detectChanges();
 
-      const signOutElement = fixture.debugElement.queryAll(By.css('#header-logout-link'));
-      expect(signOutElement.length).toBeGreaterThan(0);
-      const el = signOutElement[0].nativeElement as HTMLElement;
-      expect(el.innerHTML).toContain('Sign out');
+        const href = fixture.debugElement.query(By.css('#header-logout-link')).nativeElement.getAttribute('href');
+        expect(href).toEqual(`/${component.logoutRoute}`);
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts]: Re add the removed test but with a more accurate name showing it checks if the logout link is present when logged in. And one that checks that it is not present when not logged in.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts [hunk 5]
```diff
@@ -2,17 +2,14 @@ import { Component, Input } from '@angular/core';
...
+    @Input() loggedIn: boolean;
 
-  constructor(private router: Router) {
-  }
+    logoutRoute = 'logout';
 
-  logout() {
-    this.router.navigate(['/logout']);
-  }
+    constructor() {}
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts:14]: Remove empty constructor.

**Review by alberto-hortiguela — DISMISSED** [review]
LGTM

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts [hunk 6]
```diff
@@ -2,17 +2,12 @@ import { Component, Input } from '@angular/core';
 import { Router } from '@angular/router';
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts:2]: Remove this line as it is flagged as a code smell in sonar cloud.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts [hunk 7]
```diff
@@ -1,64 +1,58 @@
...
 
-      const link = fixture.debugElement.query(By.css('#header-logout-link'));
-      link.triggerEventHandler('click', null);
+    it('logout button is not rendered when user is not logged in', () => {
+        component.loggedIn = false;
+        fixture.detectChanges();
 
-      expect(router.navigate).toHaveBeenCalledWith(['/logout']);
+        const logoutButton = fixture.debugElement.query(By.css('#header-logout-link'));
+        expect(logoutButton).toBeNull();
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:56]: Add a test that checks it is present when logged in is true.

Comment (omio-ma) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:56]: Hi Farron, that edge case is covered by the other test

**Review by alberto-hortiguela — APPROVED** [review]
LGTM

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts (hunk 7 above)
Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:56]: I see now. I think it would be best if you reverted the test for the click event to how it was before, rather than testing the href and have a test for the logout button being present just to be more specific.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts [hunk 8]
```diff
@@ -1,18 +1,12 @@
...
-  @Input() loggedIn: boolean;
+    @Input() loggedIn: boolean;
 
-  constructor(private router: Router) {
-  }
-
-  logout() {
-    this.router.navigate(['/logout']);
-  }
+    logoutRoute = 'logout';
```
Comment (FarronF) [inline ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.ts:11]: Use the paths.ts file.

File: ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts (hunk 7 above)
Comment (omio-ma) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:56]: I don't think it would be possible as router is no longer being injected through the constructor so I cant pass a mocked router and then check if its being called, initially thats what I was trying to do, but the problem here is that the router link is only referenced in the html template file unless I am missing something like a RouterTestingModule library for html templates?

Comment (FarronF) [reply ServiceWebsite/ServiceWebsite/ClientApp/src/app/modules/shared/header/header.component.spec.ts:56]: Ah yeah, makes sense. I didn't consider router being removed. Probably could use RouterTestingModule to be completely thorough but that's fine.

**Review by FarronF — APPROVED** [review]

**Review by alberto-hortiguela — APPROVED** [review]
LGTM
